### DISCUSSION TITLE
Read Characteristic using IO programming model?

### DISCUSSION THREAD

**Discussion by Jesus805** [bluez/bluer#72] [Q&A] [answered]
I noticed the only example for a read characteristic is in [gatt_server_cb.rs](https://github.com/bluez/bluer/blob/8ffd4aeef3f8ab0d65dca66eb5a03f223351f586/bluer/examples/gatt_server_cb.rs#L50). How do we use Read Characteristics in the IO programming model? Like [here](https://github.com/bluez/bluer/blob/8ffd4aeef3f8ab0d65dca66eb5a03f223351f586/bluer/examples/gatt_server_io.rs)?

**Comment by Jesus805** [comment, marked as answer]
Nevermind, it seems this library only supports CharacteristicRead callbacks

**Reply by surban** [reply]
Yes, this is a limitation of the BlueZ API.

To provide an IO-based read characteristic we would need an `AcquireRead` function, which is not provided by BlueZ over D-Bus. See [here](https://github.com/bluez/bluez/blob/338ba4352b6cadafe61808c21302c0abe2168e02/doc/gatt-api.txt#L119) for reference.

However, you can achieve similar results by using notify characteristic in the IO programming model, provided that your Bluetooth device supports it.
